### PR TITLE
Add support for HTTP request body as file

### LINKED ISSUE #391: Feature request: ability to specify HTTP probe POST body content as on-disk file
Blackbox exporter config allows specifying POST body for HTTP probe as an in-line string in the config file. I have a circumstance where I want to probe with some large body sizes, and there is a significant memory impact of treating those bodies as inline strings.

I propose adding a `BodyPath` parameter to the HTTP prober, that will be used to open a file on disk and use the content of that file as the body of the probe's requests. Something like this:

```
type HTTPProbe struct {
	// Defaults to 2xx.
	ValidStatusCodes       []int             `yaml:"valid_status_codes,omitempty"`
...
	Body                   string            `yaml:"body,omitempty"`
	BodyPath               string            `yaml:"body_path,omitempty"`
	HTTPClientConfig       config.HTTPClientConfig `yaml:"http_client_config,inline"`
}
```

### PR DESCRIPTION
**Pull request by danteu**
Resolves #391

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -330,6 +331,10 @@ func (s *HTTPProbe) UnmarshalYAML(unmarshal func(interface{}) error) error {
 		s.HTTPClientConfig.FollowRedirects = !*s.NoFollowRedirects
 	}
 
+	if s.Body != "" && s.BodyFile != "" {
+		return errors.New("body and body_file must not both be set")
```
Comment (electron0zero) [inline config/config.go:335]:
```suggestion
		return errors.New("setting body and body_file both are not allowed")
```

File: CONFIGURATION.md [hunk 2]
```diff
@@ -129,7 +129,11 @@ The other placeholders are specified separately.
   [ ip_protocol_fallback: <boolean> | default = true ]
 
   # The body of the HTTP request used in probe.
-  body: [ <string> ]
+  [ body: <string> ]
+
+  # Read the HTTP request body from from a file.
+  # It is mutually exclusive with `body`.
+  [ body_file: <filename> ]
```
Comment (electron0zero) [inline CONFIGURATION.md:136]: can we update examples as well

File: prober/http_test.go [hunk 3]
```diff
@@ -1405,3 +1406,64 @@ func TestSkipResolvePhase(t *testing.T) {
...
+		}
+	}))
+	defer ts.Close()
+
+	registry := prometheus.NewRegistry()
+	testCTX, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	result := ProbeHTTP(testCTX, ts.URL, config.Module{Timeout: time.Second, HTTP: config.HTTPProbe{
+		IPProtocolFallback: true,
+		BodyFile:           tmpBodyFile.Name(),
```
[the remark this answers is not in the document]
Comment (danteu) [reply prober/http_test.go]: done

File: config/config.go (hunk 1 above)
Comment (danteu) [reply config/config.go]: done

File: CONFIGURATION.md (hunk 2 above)
Comment (danteu) [reply CONFIGURATION.md:136]: done

**Review by electron0zero — APPROVED** [review]
lgtm :+1:, thanks for the pr 🎊

**Comment by danteu** [comment]
Hi, I was wondering if this could be merged soon or if other changes are required.
ping @roidelapluie (I think you are assigned as the second reviewer)

File: prober/http_test.go [hunk 4]
```diff
@@ -22,6 +22,7 @@ import (
 	"crypto/x509"
 	"encoding/pem"
 	"fmt"
+	"io/ioutil"
```
Comment (mem) [inline prober/http_test.go:25]:
```suggestion
	"io"
```

File: prober/http_test.go [hunk 5]
```diff
@@ -1405,3 +1406,51 @@ func TestSkipResolvePhase(t *testing.T) {
...
+	}
+
+	tests := []config.HTTPProbe{
+		{IPProtocolFallback: true, Body: body},
+		{IPProtocolFallback: true, BodyFile: tmpBodyFile.Name()},
+	}
+
+	for i, test := range tests {
+		ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			b, err := ioutil.ReadAll(r.Body)
```
Comment (mem) [inline prober/http_test.go:1430]:
```suggestion
			b, err := io.ReadAll(r.Body)
```

**Review by mem — COMMENTED** [review]
A couple of minor changes. ioutil is deprecated and go.mod already lists version 1.18 as expected.

**Review by mem — APPROVED** [review]

**Comment by lefuturiste** [comment]
Hi! I'm also interested in this feature, what is preventing this to go forward to be merged?

**Pull request merged**